nifi-nar-bundles/nifi-salesforce-bundle/nifi-salesforce-processors/src/main/java/org/apache/nifi/processors/salesforce/QuerySalesforceObject.java

```diff
@@ -385,144 +354,181 @@ private void processQuery(ProcessContext context, ProcessSession session) {
             try (InputStream response = getResultInputStream(nextRecordsUrl.get(), customQuery)) {
-                if (originalFlowFile != null) {
-                    outgoingFlowFile = session.create(originalFlowFile);
-                } else {
-                    outgoingFlowFile = session.create();
-                }
+                FlowFile outgoingFlowFile = createOutgoingFlowFile(session, originalFlowFile);
                 outgoingFlowFiles.add(outgoingFlowFile);
-                outgoingFlowFile = session.write(outgoingFlowFile, parseHttpResponse(response, nextRecordsUrl, totalSize));
-                int recordCount = nextRecordsUrl.get() != null ? 2000 : Integer.parseInt(totalSize.get()) % 2000;
+                outgoingFlowFile = session.write(outgoingFlowFile, parseCustomQueryResponse(response, nextRecordsUrl, totalSize));
+                int recordCount = nextRecordsUrl.get() != null ? MAX_RECORD_COUNT : Integer.parseInt(totalSize.get()) % MAX_RECORD_COUNT;
                 Map<String, String> attributes = new HashMap<>();
                 attributes.put(CoreAttributes.MIME_TYPE.key(), "application/json");
-                attributes.put(TOTAL_RECORD_COUNT, String.valueOf(recordCount));
+                attributes.put(TOTAL_RECORD_COUNT_ATTRIBUTE, String.valueOf(recordCount));
                 session.adjustCounter("Salesforce records processed", recordCount, false);
                 session.putAllAttributes(outgoingFlowFile, attributes);
             } catch (IOException e) {
                 throw new ProcessException("Couldn't get Salesforce records", e);
             } catch (Exception e) {
-                if (originalFlowFile != null) {
-                    session.transfer(originalFlowFile, REL_FAILURE);
-                    isOriginalTransferred = true;
-                }
-                getLogger().error("Couldn't get Salesforce records", e);
-                session.remove(outgoingFlowFiles);
-                outgoingFlowFiles.clear();
+                handleError(session, originalFlowFile, isOriginalTransferred, outgoingFlowFiles, e, "Couldn't get Salesforce records");
                 break;
             }
         } while (nextRecordsUrl.get() != null);
 
+        transferFlowFiles(session, outgoingFlowFiles, originalFlowFile, isOriginalTransferred, startNanos, "custom");
+    }
+
+    private void transferFlowFiles(ProcessSession session, List<FlowFile> outgoingFlowFiles, FlowFile originalFlowFile, AtomicBoolean isOriginalTransferred,
+                                   long startNanos, String urlDetail) {
         if (!outgoingFlowFiles.isEmpty()) {
             session.transfer(outgoingFlowFiles, REL_SUCCESS);
+            long transferMillis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startNanos);
+
+            outgoingFlowFiles.forEach(ff ->
+                    session.getProvenanceReporter().receive(ff, salesforceRestService.getVersionedBaseUrl() + urlDetail, transferMillis)
```

Comment:
Wrong URi is created for custom query, a / is missing:

"https://xxx.my.salesforce.com/services/data/v54.0custom"